foundry/foundry.py

```diff
@@ -365,12 +392,201 @@ def publish(self, foundry_metadata, data_source, title, authors, update=False,
+            # update destination path if we have walked into a child directory
+            if root is not local_data_path:
+                # get the child directory relative path
+                subpath = os.path.relpath(root, local_data_path)
+                # update destination path to include child directories (ie subpaths)
+                dest_path = os.path.join(parent_dest_path, subpath)
+            # create child directories on endpoint
+                try:
+                    self.transfer_client.operation_mkdir(endpoint_id=endpoint_id, path=dest_path)
+                except TransferAPIError as e:
+                    raise IOError(f"Error while creating child directory {dest_path}: {e.message}") from e
+            # get local path to file to upload
+            for filename in files:
+                filepath = os.path.join(root, filename)
+                # upload file to destination path on endpoint
+                result = self._upload_file(filepath, https_base_url, dest_path, endpoint_id)
+                results.append(result)
+        return results
+
+    def _upload_file(self, filepath, https_base_url, dest_path, endpoint_id):
+        """Upload an individual file to a Globus endpoint using HTTPS PUT
+        Args:
+            filepath (str): The path to the local file to upload.
+            https_base_url (str): The URL for a given Globus endpoint.
+            dest_path (str): The path to the folder to be written to on the given endpoint.
+            endpoint_id (str): The UUID designating the exact Globus endpoint. Can be obtained via the Globus Web UI or
+                the SDK. This must be the same endpoint pointed to by the https_base_url.
+        Returns
+        -------
+            (dict): The HTTPS response dict from a PUT request
+        """
+        # lets you HTTPS to specific endpoint (NCSA endpoint by default)
+        scope = f"https://auth.globus.org/scopes/{endpoint_id}/https"
+        # Get the authorization header token (string for the headers dict) HTTPS upload
+        auth_gcs = AuthClient(authorizer=self.auths[scope])
+        header = auth_gcs.authorizer.get_authorization_header()
+
+        # get Globus endpoint path to write to
+        filename = os.path.split(filepath)[1]
+        # need to strip out leading "/" in dest_path for join to work
+        endpoint_dest = os.path.join(https_base_url, dest_path.lstrip("/"), filename)
+
+        # upload via HTTPS
+        with open(filepath, 'rb') as f:
+            reply = requests.put(
+                endpoint_dest,
+                headers={"Authorization": header},
+                files={
+                    'file': (filename, f, 'application/octet-stream')
+                }
+            )
+        if reply.status_code != 200:
+            raise IOError(f"Error on HTTPS PUT, got response {reply.status_code}: {reply.text}")
+        # Return the response
+        return reply
+
+    def make_globus_link(self, endpoint_id, path):
```

Comment:
How do you feel about moving these functions out into an "upload" submodule? I'm hoping to cut down on the number of functions of the base object to make it simpler to understand how to use

Reply:
I feel great about that! There's certainly a lot of helper functions needed for this